pkg/kube/controllers/extendedjob/job_create_secret.go

```diff
@@ -0,0 +1,286 @@
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// PersistOutput converts the output files of each container
+// in the pod related to an ejob into a kubernetes secret.
+func PersistOutput(namespace string) error {
+
+	// hostname of the container is the pod name in kubernetes
+	podName, err := os.Hostname()
+	if err != nil {
+		return errors.Wrapf(err, "Failed to fetch pod name.")
+	}
+	if podName == "" {
+		return errors.Wrapf(err, "Pod name is empty.")
+	}
+
+	// Authenticate with the cluster
+	clientSet, versionedClientSet, err := authenticateInCluster()
+	if err != nil {
+		return err
+	}
+
+	// Fetch the pod
+	pod, err := clientSet.CoreV1().Pods(namespace).Get(podName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to fetch pod %s", podName)
+	}
+
+	// Fetch the exjob
+	exjobName := pod.GetLabels()["fissile.cloudfoundry.org/ejob-name"]
+
+	exjobClient := versionedClientSet.ExtendedjobV1alpha1().ExtendedJobs(namespace)
+	exjob, err := exjobClient.Get(exjobName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to fetch exjob")
+	}
+
+	// convert output if needed
+	if exjob.Spec.Output != nil {
+		err = convertOutputToSecret(pod, exjob.Spec.Output.NamePrefix, namespace, clientSet, exjob)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func convertOutputToSecret(pod *corev1.Pod, namePrefix string, namespace string, clientSet *kubernetes.Clientset, exjob *ejv1.ExtendedJob) error {
+	fileNotifyChannel := make(chan int)
+	errorChannel := make(chan error)
+
+	// Loop over containers and create secrets for each output file for container
+	for containerIndex, container := range pod.Spec.Containers {
+
+		if container.Name == "output-persist" {
+			continue
+		}
+
+		filePath := fmt.Sprintf("%s%s%s", "/mnt/quarks/", container.Name, "/output.json")
+
+		// Go routine to wait for the file to be created
+		go waitForFile(containerIndex, filePath, fileNotifyChannel, errorChannel)
+	}
+
+	// wait for all the go routines
+	for i := 0; i < len(pod.Spec.Containers)-1; i++ {
+		select {
+		case containerIndex := <-fileNotifyChannel:
+			outputContainer := pod.Spec.Containers[containerIndex]
+			err := createOutputSecret(outputContainer, namePrefix, namespace, pod.Name, clientSet, exjob)
+			if err != nil {
+				return err
+			}
+		case failure := <-errorChannel:
+			return errors.Wrapf(failure, "Failure waiting for output file for container in pod %s", pod.GetName())
+		}
+	}
+	return nil
+}
+
+// waitForFile waits for the file to be created
+func waitForFile(containerIndex int, filePath string, fileNotifyChannel chan<- int, errorChannel chan<- error) {
+
+	for {
+		time.Sleep(5 * time.Second)
```

Comment:
we should use https://github.com/fsnotify/fsnotify instead

Reply:
Done.